dask_jobqueue/core.py

```diff
@@ -1,20 +1,95 @@
+from __future__ import absolute_import, division, print_function
+
 import logging
+import math
 import shlex
 import socket
 import subprocess
 import sys
+from collections import OrderedDict
 from contextlib import contextmanager
 
 import dask
 import docrep
 from distributed import LocalCluster
 from distributed.deploy import Cluster
-from distributed.utils import get_ip_interface, ignoring, parse_bytes, tmpfile
+from distributed.utils import get_ip_interface, parse_bytes, tmpfile
+from distributed.diagnostics.plugin import SchedulerPlugin
+from sortedcontainers import SortedDict
 
 logger = logging.getLogger(__name__)
 docstrings = docrep.DocstringProcessor()
 
 
+def _job_id_from_worker_name(name):
+    ''' utility to parse the job ID from the worker name
+
+    template: 'prefix-jobid[-proc]'
+    '''
+    pieces = name.split('-')
+    if len(pieces) == 2:
+        return pieces[-1]
+    else:
+        return pieces[-2]
+
+
+class Job(object):
+    def __init__(self, job_id, status=None):
+        self.job_id = job_id
+        self.status = status
+        self.workers = SortedDict()
+
+    def update(self, worker=None, status=None):
+        ''' update the status this job'''
+        if worker is not None:
+            self.workers[worker.address] = worker
+        if status is not None:
+            self.status = status
+
+    def __repr__(self):
+        return "<%s: %r, status: %r, workers: %r>" % (
+            self.__class__.__name__, self.job_id, self.status, self.workers)
```

Comment:
Thoughts on removing this class and just using dictionaries of sets in the plugin?  It seems like we have all of this information implicit in the `*_jobs` attributes already.

```python
self.STATUS_jobs[JOB_ID] = {WORKERS}
```

Reply:
Yes, I think we could do this. In an early version of this work, I had some additional functionality on this class but that has all been removed now.